consensus/src/da.rs

```diff
@@ -0,0 +1,269 @@
+            for (_hash, txn) in unclaimed_txns {
+                txns.push(txn.clone());
+            }
+            return Some(txns);
+        }
+        None
+    }
+    /// Run one view of the DA leader task
+    #[instrument(skip(self), fields(id = self.id, view = *self.cur_view), name = "Sequencing DALeader Task", level = "error")]
+    pub async fn run_view(self) -> Option<DACertificate<TYPES>> {
+        // Prepare teh DA Proposal
+        let parent_leaf = if let Some(parent) = self.parent_leaf().await {
+            parent
+        } else {
+            warn!("Couldn't find high QC parent in state map.");
+            return None;
+        };
+        let starting_state = if let Left(state) = &parent_leaf.state {
+            state
+        } else {
+            warn!("Don't have last state on parent leaf");
+            return None;
+        };
+        let mut block = starting_state.next_block();
+        let txns = self.wait_for_transactions().await?;
+
+        for txn in txns {
+            let new_block_check = block.add_transaction_raw(&txn);
+            // TODO: We probably don't need this check her or replace with "structural validate"
+            if let Ok(new_block) = new_block_check {
+                if starting_state.validate_block(&new_block, &self.cur_view) {
+                    block = new_block;
+                    continue;
+                }
+            }
+        }
+        let block_commitment = block.commit();
+
+        if let Ok(_new_state) = starting_state.append(&block, &self.cur_view) {
+            let consensus = self.consensus.read().await;
+            let signature = self.api.sign_da_proposal(&block.commit());
+            let leaf: DAProposal<TYPES, ELECTION> = DAProposal {
+                deltas: block,
+                view_number: self.cur_view,
+                _pd: PhantomData,
+            };
+            let message =
+                ConsensusMessage::<TYPES, DALeaf<TYPES>, DAProposal<TYPES, ELECTION>>::Proposal(
+                    Proposal { leaf, signature },
+                );
+            // Brodcast DA proposal
+            if let Err(e) = self.api.send_broadcast_message(message.clone()).await {
+                consensus.metrics.failed_to_send_messages.add(1);
+                warn!(?message, ?e, "Could not broadcast leader proposal");
+            } else {
+                consensus.metrics.outgoing_broadcast_messages.add(1);
+            }
+        } else {
+            error!("Could not append state in high qc for proposal. Failed to send out proposal.");
+        }
+
+        // Wait for DA votes or Timeout
+        let lock = self.vote_collection_chan.lock().await;
+        let mut vote_outcomes = HashMap::new();
+        let threshold = self.api.threshold();
+        let mut stake_casted = 0;
+
+        while let Ok(msg) = lock.recv().await {
+            if Into::<ConsensusMessage<_, _, _>>::into(msg.clone()).view_number() != self.cur_view {
```

Comment:
Just noticed that we had this check before validating the vote, so it actually doesn't matter whether we use `self.cur_view` (as in the current version) or `vote.view_number()` (as in your previous implementation) for validation.

I'll add this check to `next_leader.rs` as well, but the DA logic looks good!